workflow/controller/operator.go

```diff
@@ -421,9 +421,12 @@ func (woc *wfOperationCtx) operate(ctx context.Context) {
 			default:
 				if !errorsutil.IsTransientErr(err) && !woc.wf.Status.Phase.Completed() && os.Getenv("BUBBLE_ENTRY_TEMPLATE_ERR") != "false" {
 					woc.markWorkflowError(ctx, x)
-					return
 				}
 			}
+			err = woc.deletePVCs(ctx)
+			if err != nil {
+				woc.log.WithError(err).Warn("failed to delete PVCs")
+			}
```

Comment:
Do you know why this failed when using `workflowTemplateRef` but not when the whole workflow was inlined? I'm wondering if there is a deeper problem to be fixed rather than adding another deletePVC call

Reply:
Yeah, so I think the root issue is that before the controller creates PVCs, when it does validation for `Workflow`s, there are 2 cases:
- If the `Workflow` doesn't refer to any `WorkflowTemplate` and it is inlined, controller will validate the workflow itself (inlined workflow), this is fine since the bad onExit templates will not pass the validation, and PVC will not even be created.
- If the `Workflow` is referred to a `WorkflowTemplate`, the controller will also only validate the workflow, but not the templates from `WorkflowTemplate` spec. (https://github.com/argoproj/argo-workflows/blob/master/workflow/controller/operator.go#L3589). In this case, PVC will be created, and never be deleted. That's why we have different behavior.

I will open a PR to fix the above issue. However, this  deletePVC call is still needed since it could be other errors when onExit executes(not only because onExit bad templates).

Reply:
> I will open a PR to fix the above issue since I feel like it is an enhancement and bugfix for workflow validation

Yes, I agree, seems with using `workflowTemplateRef`, we do less validation than when they are inlined.

> Also, this deletePVC call is still needed since it could be other errors when onExit executes(not only because onExit bad templates).

I originally had the same thought as you and felt that this was a bug with cleaning PVCs when onExit has errors. However, when I tried the following workflow, it creates and then properly deletes the PVC when onExit fails.

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: wft-pvc-
spec:
  volumeClaimTemplates:
    - metadata:
        name: builddir
      spec:
        accessModes: [ "ReadWriteMany" ]
        resources:
          requests:
            storage: 1Mi

  volumeClaimGC:
    strategy: OnWorkflowCompletion

  entrypoint: whalesay
  onExit: notify

  templates:
  - name: whalesay
    container:
      image: docker/whalesay:latest
      command: [cowsay]
      args: ["hello world"]

  - name: notify
    container:
      image: docker/whalesay:latest
      command: [sh, -c]
      args: ["exit 1"]
```

This tells me that probably onExit PVC cleanup is working properly, and we only need the extra validation that is missing when using `workflowTemplateRef`. In other words, I think this PR may not be necessary if we fix the validation.

Reply:
Yeah, the above workflow works since the `Status` of the workflow is `Failed` not `Error`, I am trying to construct a workflow that will cause an error here and return here(this is the same place that the workflow that uses `workflowTemplateRef` will return): https://github.com/argoproj/argo-workflows/blob/master/workflow/controller/operator.go#L427, and I don't know how I can do that. Maybe you have an idea? If we can construct a workflow return here, then I guess the PVCs will not be deleted?

Reply:
I have added the workflow validating logic in this commit: https://github.com/argoproj/argo-workflows/pull/10424/commits/d18db259ff617909ca1ff24bfb72e994024befc0, and I have tested without the `deletePVC` call added by this PR, and it works, and now the workflow will fail to pass the validation, and PVCs will not be created.
If you think what I mentioned in the last comment is not an issue, I will just remove the `deletePVC` call and change the PR title and description.